### DISCUSSION TITLE
Multitenant support

### DISCUSSION THREAD

**Discussion by bharatnpti** [quarkusio/quarkus#29806] [Community]
I want to add support of multitenant in different modules. I would like to start with Rest Clients (@RegisterRestClient).

The config would be in below structure where a single client can have different configurations for different tenants.

quarkus.rest-client.firstRestClient.**tenant1**.url=https://quarkus-tenant1-firstrestclient.free.beeceptor.com
quarkus.rest-client.firstRestClient.**tenant2**.url=https://quarkus-tenant2-firstrestclient.free.beeceptor.com

**Code approach** -

resteasy-classic contains rest-client and rest-client-config.

Introduce new module rest-client-multiteant-config which contains the config class to read the properties/yaml as shown above.

However, presently rest-client and rest-client-config are tightly coupled as rest-client directly uses RestClientsConfig class from rest-client-config.

I propose to have a module containing only the abstraction and adding it as a dependency to rest-client-internal.
Then we can have separate modules :-

1) rest-client which has rest-client-internal and rest-client-config as dependency for supporting non-multitenant use case
2) rest-client-multitenant which has rest-client-internal and rest-client-multitenant-config as dependencies to support multitenant support.

**Comment by bharatnpti** [comment]
@TomasHofman - can you please provide inputs

**Reply by TomasHofman** [reply]
Hello @bharatnpti , can you describe more what / who is a "tenant" in the use case you have in mind?

**Comment by bharatnpti** [comment]
Hi @TomasHofman,

Below scenario demonstrates the meaning of tenant :-

Organization A - This org has a microservice(let's call it **aggregator microservice**) which fetches data from different clients/organization. The format of data,request, reponse etc is same. What differs is the base url on which these different clients exposes the api for retrieving the data.

**Tenat** 1 - Organization 1 - url - http://**abc.com**/data/
**Tenat** 2 - Organization 2 - url - http://**xyz.com**/data/

Now the aggregator microservice uses the @RegisterRestClient to make call to Organization 1 and Organization 2 api.

Presently, there is no way to specify different url for same @RegisterRestClient interface.

I would like to add the capability to RestClient to have different url configured in yml/properties for different tenant

**Reply by jamezp** [reply]
How would the client know which URL it needs to invoke?

**Reply by bharatnpti** [reply]
It would be based on the tenant which would be set at the time of request/process initiation.
The tenant would be present in ThreadLocal for non-reactive rest client.

Use Case :- 

- Tenant 1 makes a request to aggregator microservice passing the tenant value in request(headers).
- A interceptor will read the value from the request and populate in ThreadLocal.
- Now aggregator microservice needs to call api of one the tenant, the config will be returned for the tenant present in ThreadLocal.
- Using the config returned the request will created and finally calls the api of required tenant.

**Reply by jamezp** [reply]
Why not just use the `RestClientBuilder` then? You could easily do that the way you describe:
```java
final ExampleClient client = RestClientBuilder.newBuilder()
    .baseUri(determineUri(request))
    .build(ExampleClient.class);
```

**Reply by bharatnpti** [reply]
Even presently without using @RegisterRestClient annotation user can use RestClientBuilder.
So, I think the reason would be same as of using @RegisterRestClient instead of RestClientBuilder (in present scenario of non multi-tenant environment) which are :-

1) It is more convenient to simply use the annotation than to create client using RestClientBuilder.
2) No need to write code for creating the clients only interface and annotations required.
3) CDI injection support

Basically, to have the **benefit of declarative web service client**

Also, there would be consistency from the point of view developer when using RestClient of Quarkus whether they have a multi-tenant environment or not.

I am open to both approaches, however I think supporting multi-tenancy for @RegisterRestClient will be better.


Would love hear your thoughts on same

**Reply by jamezp** [reply]
I don't even see how it could work. How would it know which URL/URI to use? There would have to be some kind of CDI qualifier and even then I'm not sure how it would work. You could make your own CDI producer which would handle the `determinUri(request)`, but I'm not sure how an implementation would be able to just know which one you'd want to use.

**Reply by bharatnpti** [reply]
I do have a high level idea on how to achieve this.
I can give it a try if we are open to the idea of adding support of multi-tenant to @RegisterRestClient

**Reply by jamezp** [reply]
Where do these projects live? For RESTEasy Classic the project is here https://github.com/resteasy/resteasy-microprofile. The config and rest-client implementation don't really have anything to do with each other.

**Reply by bharatnpti** [reply]
I am talking about rest-client and rest-client-config present at - https://github.com/quarkusio/quarkus/tree/main/extensions/resteasy-classic

**Reply by jamezp** [reply]
Ah. I don't really have a say on what happens there. I'm still not sure how you'd be able to decide which URL to use with no information.

**Reply by bharatnpti** [reply]
ohh ok, any idea who is the correct person for this ?

**Reply by jamezp** [reply]
I'm not too sure there.

**Reply by maxandersen** [reply]
Maybe @geoand ?

Though this kind of new features feels more like a resteasy reactive effort ?

**Reply by geoand** [reply]
I might be missing something, but isn't this what 

```java
@RegisterRestClient(configKey = "extensions-api")
````

and 

```properties
quarkus.rest-client.extensions-api.url=https://stage.code.quarkus.io/api
```

achieve?

**Reply by bharatnpti** [reply]
Hi @geoand,

It's fine for non-multitenant application, however does not have support for multitenant applications.
Please refer to https://github.com/quarkusio/quarkus/discussions/29806#discussioncomment-4398823 for multitenant use-case

**Reply by geoand** [reply]
Wouldn't it be far easier to just have a base interface and extend that one with the specific use cases and annotate those?

In your proposal, what would the API look like?

**Reply by bharatnpti** [reply]
you mean extend the base interface for each tenant ?
I think approach suggested by @jamezp https://github.com/quarkusio/quarkus/discussions/29806#discussioncomment-4402055 will be much better if user of Quarkus has to do it.

However, I am proposing to add the capability in Quarkus itself.

Approach - rest-client will always the existing RestClientConfig object to build RestClient, however there will be separate modules for rest-client-config for multitenant and non-multitenant use cases which which will resolve and return RestClientConfig.

1) Decouple rest-client and rest-client-config modules by introducing abstraction.

rest-client-config-core (abstraction - contains only interface)
rest-client-config - implements rest-client-config-core and provides existing functionality
rest-client-config-multitenant - implements rest-client-config-core, reads the config in multi-tenant env and return the RestClientConfig instance (basically changing the structure of RestClient**s**Config to have Map<String, Map<String, RestClientConfig> instead of Map<String, RestClientConfig>)

existing - 
Map<String, RestClientConfig> is map of rest client name and corresponding configuration (RestClientConfig)

new - 
Map<String, Map<String, RestClientConfig>> will be map of rest client name and value will be map of tenant and corresponding config (RestClientConfig)


Multitenant config :-

quarkus.rest-client.myRemoteService.**tenant1**.url=
quarkus.rest-client.myRemoteService.**tenant1**.connectTimeout=

quarkus.rest-client.myRemoteService.**tenant2**.url=
quarkus.rest-client.myRemoteService.**tenant2**.connectTimeout=

rest-client-core - has core functionality (as present) but use rest-client-config-core as dependency instead of rest-client-config

2) rest-client - has rest-client-config and rest-client-core as dependency - supporting the existing functionality
3) rest-client-multitenant - has rest-client-config-multitenant and  rest-client-core as dependency - supporting the multitenant functionality

**Reply by geoand** [reply]
> I think approach suggested by @jamezp https://github.com/quarkusio/quarkus/discussions/29806#discussioncomment-4402055 will be much better if user of Quarkus has to do it.

Sure, that's possible as well

**Reply by geoand** [reply]
I personally don't see much value, but if you want to try and do a PR, that's fine. Just keep in mind that the Reactive REST Client is the one we plan to support going forward

**Reply by bharatnpti** [reply]
sure, I'll give it a try.
Yes I am aware of future support, would like to start with classic rest client then reactive one

**Reply by bharatnpti** [reply]
Hi @geoand,

I have started working on it.
I am facing one issue while decoupling the rest-client-config and rest-client which is described here. Can you please help me with it or loop in someone who might be able to help [Issue](https://github.com/quarkusio/quarkus/issues/32144)

**Comment by bharatnpti** [comment]
Hi @asoldano, @ronsigal - can you please provide inputs

**Reply by asoldano** [reply]
@jamezp ^
